### PR TITLE
Localize after cache

### PR DESCRIPTION
**Pull request by ziv17**
Updated widget creation to allow localization after retrieval from cache. A static method was added to base class `Widget` to perform the localization. Each widget subclass needs to implement it. Function is static because after cache we do not have the widget instances. The function receives the items that were generated by that widget instance, and `request_params`.

### CODE REVIEW

File: anyway/infographics_utils.py [hunk 1]
```diff
@@ -158,6 +166,7 @@ def serialize(self):
 
 class WidgetCollection:
     widgets: List[Type[Widget]] = []
+    widgets_dict: Dict[str, Widget] = {}
```
Comment (elazarg) [inline anyway/infographics_utils.py:169]: This should be `Dict[str, Type[Widget]]`.

We don't need `widgets` anymore, we can simply iterate `widgets_dict.values()` (though we can call it `widgets`.

BTW, there's no need for `WidgetCollection` if it's never instantiated. We can have `widgets_dict` as a global variable (which is what `WidgetCollection` is, effectively), no constructor, no `get()` method, just a global dictionary and a global `register_widget` function.

And if we do have the class, no need for the `get()` method - just access the variable. `get()` is only needed if it's an expensive access that we want client code to be aware of. Languages with `@property` don't need getters.

File: anyway/infographics_utils.py [hunk 2]
```diff
@@ -169,16 +178,21 @@ def get() -> List[Type[Widget]]:
...
         WidgetCollection.widgets.append(widget_class)
+        WidgetCollection.widgets_dict[widget_class.widget_id.name] = widget_class
+        logging.debug(f"register:{widget_class.widget_id.name}:{widget_class}\n")
         return widget_class
 
 
 @WidgetCollection.register
 class AccidentCountBySeverityWidget(Widget):
-    widget_id: WidgetId = dataclasses.field(init=False, default=WidgetId.accident_count_by_severity)
+    widget_id: WidgetId = WidgetId.accident_count_by_severity
```
Comment (elazarg) [inline anyway/infographics_utils.py:188]: I don't see the need for `WidgeId` anymore. The enum is effectively replaced by the widget-collection. `widget_id` is simply the name of the class. Nothing needs to be explicitly initialized or passed around.

Comment (elazarg) [reply anyway/infographics_utils.py:188]: and `name` is essentially a nicer way of saying `cls.__name__`. If you want this syntactic sugar, it can be defined once, as a trivial classmethod.

File: anyway/infographics_utils.py (hunk 1 above)
Comment (ziv17) [reply anyway/infographics_utils.py:169]: I agree `widgets` is not needed.
Regarding `WidgetCollection`, I prefer to have the class to enclose the implementation of the gathering of the widgets. (I will prefix the dict with `__`)

File: anyway/infographics_utils.py (hunk 2 above)
Comment (ziv17) [reply anyway/infographics_utils.py:188]: The names we use are lowercase with underscore word separators. Class names are Camel-case. How do you suggest to overcome this discrepancy when using `cls.__name__`?

File: anyway/infographics_utils.py (hunk 1 above)
Comment (elazarg) [reply anyway/infographics_utils.py:169]: IMO there's no need for that kind of encapsulation, and it's not standard to use it in Python (the implementation is exposed just the same with the `get()` method). It makes it look more complex than it is.

File: anyway/infographics_utils.py (hunk 2 above)
Comment (elazarg) [reply anyway/infographics_utils.py:188]: Where do we use these names Where are the created? If they originate inside the code, they can be changed; if they are not, and we rely on the to reflect variable names inside the program (the names of the enum instances) then we expose implementation detail to the outside.

I thought they come from iterating over `WidgetCollection.widgets`.

Comment (ziv17) [reply anyway/infographics_utils.py:192]: The name is a field in the widgets that the api returns. I assume its goal is to identify the widget in the UI. We can change the current names to class names (camel notation). I do not know if and to what extent it affects the UI.

File: anyway/infographics_utils.py (hunk 1 above)
Comment (ziv17) [reply anyway/infographics_utils.py]: Using a method to get a widget class from its name will allow us to change the implementation in the future with less code changes. If it is OK with you, I prefer to leave the class, if not, I will remove it.

Comment (elazarg) [reply anyway/infographics_utils.py]: This is true in Java. Python (and C#, and many other languages) has `@property` methods so the difference is invisible to the client code; the function call syntax hints the reader that this call might have a side effect (or that it is computationally heavy) - which is not the case here.

Also, everything here is intended for in-module use, so this kind of change will be easy; and if it is not, the fact that we return a mutable collection exposes much more significant implementation detail and is prone to misuse.

Of course I'm not here to dictate anything, even if I could :) I'm just saying what I think.

File: anyway/infographics_utils.py (hunk 2 above)
Comment (elazarg) [reply anyway/infographics_utils.py:192]: Regardless of whether this is camel case or snake case, it sounds like relying on implementation detail (name of the enum values, name of the class) in the API. It's not horrible but it's not ideal either.

Comment (ziv17) [reply anyway/infographics_utils.py:192]: I am setting the enum's name to the string I need. Is it different from setting a constant to the value I need?

Comment (elazarg) [reply anyway/infographics_utils.py:188]: Yes, because the enum's name is internal to the program, and referring to it is a kind of metaprogramming. Again - this is not horrible, it's just something to keep in mind.

File: anyway/infographics_utils.py [hunk 3]
```diff
@@ -1238,7 +1335,7 @@ def get_latest_accident_date(table_obj, filters):
 def generate_widgets(request_params: RequestParams, to_cache: bool = True) -> List[Widget]:
     widgets = []
     # for w in WidgetId:
-    for w in WidgetCollection.get():
+    for w in get_widget_classes():
         # widget: Optional[Widget] = create_widget(w, request_params)
         widget: Optional[Widget] = w(request_params)
```
Comment (elazarg) [inline anyway/infographics_utils.py:1340]: When is this None? Seems like `w` is `Type[Widget]` meaning this call is a constructor call that never returns None. Is it possible for `w` to be a function?

File: anyway/infographics_utils.py (hunk 2 above)
Comment (ziv17) [reply anyway/infographics_utils.py:188]: @elazarg 
We can do 
`class WidgetId(Enum):`
`         accident_count_by_severity = 'accident_count_by_severity'`
`         most_severe_accidents_table = 'most_severe_accidents_table'`

and then
`WidgetId.accident_count_by_severity.value`
This way it does not rely on implementation. Do you prefer it?

Comment (elazarg) [reply anyway/infographics_utils.py:188]: I think it's better, yes. We can equivalently do
```python
@WidgetCollection.register('accident_count_by_severity')
class AccidentCountBySeverityWidget(Widget):
    ...
```

Comment (elazarg) [reply anyway/infographics_utils.py:188]: I see you prefer enums, It's fine, It's just that the enum represents not a widget instance but a widget type, and we already have exactly that when we define a class, so it feels like redundancy to me.

Comment (elazarg) [reply anyway/infographics_utils.py:188]: The benefit in having enum is that the reader can see a list of available widgets.

Comment (ziv17) [reply anyway/infographics_utils.py]: I think I finally understand. I removed the `WidgetId` `Enum`.

File: anyway/infographics_utils.py [hunk 4]
```diff
@@ -156,29 +131,31 @@ def serialize(self):
...
-@WidgetCollection.register
+
+@register
 class AccidentCountBySeverityWidget(Widget):
-    widget_id: WidgetId = dataclasses.field(init=False, default=WidgetId.accident_count_by_severity)
+    name: str = "accident_count_by_severity"
 
     def __init__(self, request_params: RequestParams):
-        super().__init__(request_params, WidgetId.accident_count_by_severity)
+        super().__init__(request_params, type(self).name)
```
Comment (elazarg) [inline anyway/infographics_utils.py:156]: Cool. Now this parameter becomes unnecessary. Since `str` is an immutable type (and moreover, the field is not expected to change, i.e. it is `Final`), the class/static variable `name` (initialized in the class body) is a perfectly valid to access as a field - anyone can just access it using `mywidget.name`. As we've discussed, this does not break encapsulation.

```python
class A:
    name = 'unknown'  # here only to get static type checking when we use polymorphism; can also be a property

class B(A):
    name = 'john'

b = B()
print(b.name)  # perfectly legitimate
```

tl;dr: the parameter `name` and the method `get_name()` can be removed. This is not the Java way, but this is perfectly idiomatic Python.

Comment (ziv17) [reply anyway/infographics_utils.py:156]: Hi @elazarg , I understand your point and thank you for the deep reviews. At the same time I would prefer to leave this PR as it is now, and progress with other tasks. 
Is it OK with you to approved the PR?
Thanks, Ziv

Comment (elazarg) [reply anyway/infographics_utils.py:156]: Oh sorry, I was not intentionally withholding approval... I was only giving comments. Sorry!

**Review by elazarg — APPROVED** [review]
LGTM. I was too busy giving minor comments about style :)

**Comment by ziv17** [comment]
Many thanks :)
